### ISSUE TITLE
Georeference refactoring

### ISSUE THREAD

**Issue by javagl** [CesiumGS/cesium-unreal#330]
There are [several issues related to `CesiumGeoreference`](https://github.com/search?q=CesiumGeoreference+OR+Georeference+repo%3ACesiumGS%2Fcesium-unreal&type=issues), about usability, glitches, or crashes, and general ideas for improvements. Considering these issues, and a technical talk that we had today after the standup, the consensus seems to be that the handling of the `CesiumGeoreference` (or GeoRef, for short) should be reviewed and probably refactored. 

---

Summary:

<sup>Disclaimer: The following is an attempt of a *short* summary of some points of my "mental model" that I find relevant. This may be incomplete or imprecise. Feel free to add/edit clarifictions or corrections. I only started to look at this topic in more detail.</sup>

- Actors like `Cesium3DTileset` and `GlobeAwareDefaultPawn` use a GeoRef for coordinate conversions between "Unreal World", Earth-Centered-Earth-Fixed (ECEF), East-North-Up (ENU), and Latitude/Longitude/Height. The actors store a reference to a GeoRef, and use this instance for these coordinate conversions. 
- The GeoRef may be assigned to the actors in the UI. In some cases, this happens automatically - for example, when adding a new tileset to an empty level. More generally: Under some conditions, a "default" GeoRef instance is created. Usually, all actors in one level will use the same GeoRef (but not necessarily)
- The GeoRef is "observable", and can inform "Listeners" about changes. This is done via the `ICesiumGeoreferenceable` interface. This interface is implemented by the actors that hold the GeoRef. 

Some "conceptual" issues that may already be mentioned elsewhere:

- Deleting a GeoRef will break references and bring the actors into an invalid state. Even when handling that, it is not clear how the actors should behave when they don't have a valid GeoRef.
- Resetting the GeoRef of a `GlobeAwareDefaultPawn` to `None` does not work, because the property change causes `OnConstruction` to be called, where a new default GeoRef is created if the current one is `null`
- Default GeoRef instances that are created are not always added in the World Outliner
- `ACesium3DTileset::PlayMovieSequencer` always re-creates a (new, local) GeoRef, and does not use the one from the tileset
- The actors have a reference to the GeoRef, and the GeoRef has references to the actors (because they are "observers"). This is a cycle in the object graph, and might cause issues with garbage collection, or make it hard to achieve a "consistent" state in general.

---

Notes from the discussion: 

<sup>Another disclaimer: It was a vivid discussion, with many thoughts and ideas, sometimes leading to a point where we have to look at and think about several options, and read code + documentation. The following are **not** "minutes" in a strict sense, but rather some "takeaways", a bit raw, but hopefully helpful. If I missed something important, just edit/add as you wish. </sup>

Usually all actors of a level need a GeoRef. For a user that is not deeply involved in geospatial topics, this should not be a burden: The simple/default behavior should be that the GeoRef is just there, and users can use it easily. But it should still be possible to implement more complex use-cases, where users **might** want to use different GeoRefs for actors in the same level.

Aspects refering to how the GeoRef functionality should be offered:

- One option to achieve a simple usage could be a static function library, as suggested in "Move helper transform functions to static function library" ( https://github.com/CesiumGS/cesium-unreal/issues/318 ), or some "Singleton-like" object.

- Some of the conversion functions might be static, because they need GLM methods, and no explicit GeoRef. 

- But to keep in mind: Even the static functions often need an ellipsoid (which, by the way, cannot be changed until now).

- For the current solution (non-static functions), there is a convenience function `ACesiumGeoreference::GetDefaultForActor` that either picks the first existing GeoRef from the level, or creates a new one if none exists. There already have been considerations to allow more fine-grained control here, and maybe add a `ACesiumGeoreference::GetDefaultForLevel` counterpart. 

- One problem with that kind of function is that the may return a GeoRef, that may be stored by the caller, but when the user *deleted* this GeoRef, the caller will not be notified about that. 

- Maybe the GeoRef could *not* be stored in each actor individually, but rather as part of the "World Options" in the Unreal Engine (and there could even be a list of GeoRefs). (This may simplify the structure a bit, but I think it will not solve many of the existing issues, e.g. the one about deleted GeoRefs...)

A quick summary of the implementation options, roughly sorted by "static-ity": 

- Each actor could have its own GeoRef instance (the current state)
- There could be one global GeoRef instances that the actors can use (e.g. set via the World Options)
- There could be multiple global GeoRef instances. Each actor can pick a different one.
- There could be static functions that internally *use* a global GeoRef instance to do their work
- There could be static functions in the that *receive* a GeoRef instance as a parameter to do their work

--- 

Interlude about origin placement and `ICesiumGeoreferenceable`:

- There are several types of `EOriginPlacement`, namely `TrueOrigin`, `BoundingVolumeOrigin` and `CartographicOrigin`. Some of the complexity of the `ICesiumGeoreferenceable` interface stems from the fact that implementors (specifically, the tileset) may have to update the bounding volume, and thus, affect the GeoRef in `BoundingVolumeOrigin` mode. (I haven't understood all details here, but) it could be desirable to de-couple the main GeoRef functionality and this bounding volume update.

- Objects that are `ICesiumGeoreferenceable` have a `NotifyGeoreferenceUpdated()` method (with no parameters). This means that in order to react to updates, they *need* an instance of the GeoRef. It might make sense change this function so that it receives the relevant information from the GeoRef. For example, roughly (oversimplified) as in `NotifyGeoreferenceUpdated(mat4 newCoordinateTransformMatrix)`, so that the receiver can store the new matrix, and does not need the GeoRef instance any more.

---

Aside regarding the Blueprint-friendliness:

Right now, the `ACesiumGeoreference` contains several methods in two flavors: 

- An "accurate" one that operates on a (double-precision) `glm::dvec`
- An "inaccurate" one that operates on an (single-precision) `FVector`, that is exposed for Blueprints

This is "not so nice" in several ways: The duplication of the methods within a single class makes the class large. Unreal 5 is supposed to support `double` precision, so (fingers crossed), the "inaccurate" methods could become obsolete, or receive a `DVector` counterpart. One *very rough* idea of how this could be addressed was to offer the "inaccurate" GeoRef as a dedicated class. This class could just "wrap" the accurate one, and do the downcasting to `float` where necessary. 

A pseudocode sketch:
```
// The accurate version, operating on glm::dvec
class CesiumGeoreference {
    dvec3 TransformUeToLongitudeLatitudeHeight(dvec3 ue) {
        return someMatrix * ue;
    }
    ...
}

// The inaccurate one, for blueprints, operating on FVector:
class BlueprintCesiumGeoreference {

    // Internally uses the accurate one:
    private CesiumGeoreference cesiumGeoreference;

    // Offer this for the bueprint:
    UFUNCTION(BlueprintCallable); 
    Fvector TransformUeToLongitudeLatitudeHeight(FVector ue) {
        // Upcast
        glm::dvec3 vec = createFrom(ue);
        // Call accurate one
        cesiumGeoreference.TransformUeToLongitudeLatitudeHeight(vec);
        // Downcast:
        return FVector(vec);
    }
}
```

(And when Unreal 5 comes out, we could just wrap our `CesiumGeoreference`, *without any modification*, into a new `AccurateBlueprintCesiumGeoreference` that operates on `DVector` ... in the best case).

**Comment by xuelongmu** [comment]
Fantastic summary of our discussion @javagl, thank you for putting this together.

Just a couple additional thoughts:

> * Objects that are `ICesiumGeoreferenceable` have a `NotifyGeoreferenceUpdated()` method (with no parameters). This means that in order to react to updates, they _need_ an instance of the GeoRef. It might make sense change this function so that it receives the relevant information from the GeoRef. For example, roughly (oversimplified) as in `NotifyGeoreferenceUpdated(mat4 newCoordinateTransformMatrix)`, so that the receiver can store the new matrix, and does not need the GeoRef instance any more.

Have we considered using delegates for NotifyGeoreferenceUpdated()? Currently the Georeference (according to our discussion) keeps a list of all actors that reference the GeoRef in order to call this function. Does the GeoRef need to know about all these actors? Using delegates would invert this dependency.

> For the current solution (non-static functions), there is a convenience function ACesiumGeoreference::GetDefaultForActor that either picks the first existing GeoRef from the level, or creates a new one if none exists. There already have been considerations to allow more fine-grained control here, and maybe add a ACesiumGeoreference::GetDefaultForLevel counterpart.

I was thinking perhaps we could use tags to differentiate between the "default" georeference, which would have a "DefaultGeoRef" tag, and any additional custom ones. We could then use the [`GetAllActorsOfClassWithTag` (quite fast) ](https://forums.unrealengine.com/t/get-all-actors-of-class-vs-get-all-actors-with-tag/155107/4) to retrieve the desired georeference in the static function library. When the default georef is deleted and can no longer be found, another one in the scene could be found, or if there are no others, than a new one could be created, and designated as the default.

Additionally, for any actors that still need to store their own GeoRef as a variable, the GeoRef could notify them using an interface function (existing way) or delegate (my suggestion) that's triggered on [EndPlay](https://www.brightdevelopers.com/unreal-engine-4-game-flow-actor-lifecycle-overview/) so the actors can appropriately react to this event.

**Comment by javagl** [comment]
> Have we considered using delegates for NotifyGeoreferenceUpdated()? Currently the Georeference (according to our discussion) keeps a list of all actors that reference the GeoRef in order to call this function. Does the GeoRef need to know about all these actors? Using delegates would invert this dependency.

I'm not entirely sure whether I understood that correctly. Would that roughly mean that we declare a type like

    DECLARE_DELEGATE(FGeoRefUpdateDelegate);

and then bind them with something like

    FGeoRefUpdateDelegate UpdateTileset;
    UpdateTileset.BindSP(tileset, &Tileset::DoThisUpdate);

    FGeoRefUpdateDelegate UpdatePawn;
    UpdatePawn.BindSP(pawn, &Pawn::DoThatUpdate);

and then store these delegates in the GeoRef (instead of the actors themself)? 
(I probably misunderstood that, because I don't see where this would invert the dependency)

---

Right now, the GeoRef stores the actors, and uses them in `UpdateGeoreference` for two things:

- It fetches the bounding volumes of all actors, for the `BoundingVolumeOrigin` placement mode
- It informs all actors by calling their `NotifyGeoreferenceUpdated`

The bounding volume part is only relevant in the `EOriginPlacement::BoundingVolumeOrigin` mode, and only for the `Cesium3DTileset`. In fact, the other implementations have some dummy `false`/`nullopt` implemementations for the bounding volume methods.

Warning, brainstorming and pesudocode ahead: 

The main thing that is done in `UpdateGeoreference` is essentially "computing a point". Depending on the placement mode, this is either the origin, or the bounding volume origin, or the point that is defined via lat/lon/height+Ellipsoid. The GeoRef could just have "a thing" that provides that point:

    class PointProvider {
        virtual vec3 getPoint();
    }
    class Georeference {
        PointProvider pointProvider;

        void updateGeoreference() {
            vec3 center = pointProvider->getPoint();
            ...
        }

We could then have some `PointProvider` implementations, like a `CartographicPointProvider` that computes the point based on the lat/lon/height, or a `BoundingBoxOriginPointProvider` that stores a list of tilesets and computes the point as the origin of their bounding volumes. (Maybe that looks a bit over-engineered, but ... the GeoRef would then no longer need the tilesets for the bounding volume computations).

Some caveats here: 
- The tileset has to trigger an update when its bounding volume becomes known (asynchronously). 
- The GeoRef has a relationship to `SunSky`, which uses the Latitude/Longitute (even if the `EOriginPlacement` mode is **not** `CartographicOrigin`). 
- There is this "origin rebasing" that I haven't understood this in all depth yet... 

---

> We could then use the GetAllActorsOfClassWithTag (quite fast) to retrieve the desired georeference in the static function library. When the default georef is deleted and can no longer be found, another one in the scene could be found, or if there are no others, than a new one could be created, and designated as the default.

I started writing down possible implementation options (on a high level), and this was one option that I also considered: Having some "global list" of GeoRefs, and making sure that this list **always** contains **at least** one element, so that any potentially invalid access could "fall back" to the one instance that is always there. 

But I'll definitely have to read more of the existing code and the Unreal docs to better understand how this could be implemented sensibly. When you mention "tags", I'll have to read more about that as well. But to get an idea: What would the code for calling a certain conversion function (from a user perspective) then look like? For example, right now, the conversion from lat/lon/height to UE coordinates can be done with
```
glm::dvec3 dUe = this->Georeference->TransformLongitudeLatitudeHeightToUe(dLonLatHeight);
```
With a purely static function (to avoid the GeoRef-reference headaches), it would be
```
glm::dvec3 dUe = UCesiumGeoreferenceFunctionLibrary::TransformLongitudeLatitudeHeightToUe(dLonLatHeight)
```
But how would a user select (at this point) which GeoRef should be used? Would that be something like
```
glm::dvec3 dUe = UCesiumGeoreferenceFunctionLibrary::TransformLongitudeLatitudeHeightToUe(
    dLonLatHeight, "DefaultGeoreference")
```
with the option to call 
```
glm::dvec3 dUe = UCesiumGeoreferenceFunctionLibrary::TransformLongitudeLatitudeHeightToUe(
    dLonLatHeight, "MySpecialGeoreference")
```
to use a different one (falling back to the default when necessary)? 

**Iff** this is roughly the goal: This would mean that the actors (like the tilesets) would not have a (UI-configurable) GeoRef instance, but something like a UI-configurable `FString GeoRefTagName`, right? That sounds like a reasonable way to have 1. convenient default behavior and 2. a sensible option to have, for example, two tilesets with different GeoRefs.

**Comment by xuelongmu** [comment]
I think I may have used the term dependency inversion incorrectly here, my apologies 😅 What I mean is that we could eliminate the need for a) actors to hold GeoRef refs, and b) GeoRef to hold refs for anything other than a `Cesium3DTileset`, if we do the following:

- Conversion functions could be accessed using the static function library. We might not need to expose the FString tag parameter, and simply assume they are requesting the default georeference. All the current actors get their georef by calling `GetDefaultForActor` which pretty much makes this assumption. The rare special cases that need their own GeoReference can spawn it, maintain its reference, and call methods on it outside of the function library.
- For calling `UpdateGeoreference`, the static function library could provide a function that calls `UpdateGeoreference` on a) the default GeoRef, or b) all GeoRefs in the level. 
- For `NotifyGeoreferenceUpdated`, the notified actors can bind a callback to CesiumGeoreference like so:
On CesiumGeoreference, we would define the delegate
```
DECLARE_DYNAMIC_MULTICAST_DELEGATE(FGeoRefUpdateDelegate); //using delegates in BP requires dynamic
``` 
and declare a property
```
UPROPERTY(BlueprintAssignable)
FGeoRefUpdateDelegate GeoRefUpdated;
```
and on the notified actors (Pawn, Tileset etc.) we could do something like:
```
MyGeoRef->GeoRefUpdated.AddUObject(this, &AMyPawn::HandleGeoRefUpdate);
```
`MyGeoRef` could be provided via a static function library call that returns the GeoReference with the `DefaultGeoreference` tag. I'm not wedded to using tags, as long as we can consistently keep track of a default GeoRef for the level and react appropriately if it gets deleted.

For actors that need to provide a bounding volume, we could keep using the existing `ICesiumGeoreferenceable` interface for that. As for improvements to the bounding volume calculations with your suggestions of PointProvider implementations, I will take some time to look into that further, as I don't understand that code well enough yet 😄

**Comment by javagl** [comment]
There are many points to consider, usability (for use-cases that we might not even imagine yet) and possible technical implications. A few short comments:

---

> The rare special cases that need their own GeoReference can spawn it, maintain its reference, and call methods on it outside of the function library.

Some questions related to that (that *may* have simple anwers, or not make sense - I'm just thinking out loud here) 

- Will these spawned GeoReferences show up in the World Outliner? 
  - If so: The actors will have to handle the case that some user just hits "Delete"
  - If not: Will it be possible to let two actors have *the same* (non-default) GeoRef? (How?)

---
 
> `MyGeoRef` could be provided via a static function library call that returns the GeoReference with the `DefaultGeoreference` tag. 

(Callers may store a reference to this `MyGeoRef`. Whether or not this is an issue depends on whether (and maybe how) we make sure that it's impossible to *delete* this instance - but you already mentioned that).  

Beyond that: How would the "update notification handling" work for the **non**-default GeoRef that some actor may have spawned? **Iff** it should be possible for two actors to share a non-default GeoRef, we'd need something like

    GeoRefFromActorA->GeoRefUpdated.AddDynamic(actorB, &AMyPawn::HandleGeoRefUpdate); // OR
    GeoRefFromActorB->GeoRefUpdated.AddDynamic(actorA, &AMyPawn::HandleGeoRefUpdate);

Roughly: Who is attaching (when) to which GeoRef, and where does he get that GeoRef instance from?

---

I haven't looked at the topic of delegates in all detail yet, but based on my vague understanding, the current options are

- Let each `GeoRef` individually and *directly* store a reference to the actor (that is done right now)
- Let the "global" `GeoRef` store *delegates* (instead of actors) - this is what you suggested

The second one would in any case be desirable, to achieve some sort of "decupling": What if *something* wants to be informed about GeoRef updates, but this thing is *not* an actor that can sensibly implement `ICesiumGeoreferenceable`? We essentially already have this situation - most `ICesiumGeoreferenceable` implementations are "dummies". Boiling this down to a generic, single-method update certainly makes sense.

For the question of how the "wiring" takes place: It *might* be helpful to hide parts of that. On the one hand, it *might* be easier to use. On the other hand, it *might* allow to change things later, without breaking existing code.  I'm roughly (very roughly) thinking about hiding the *actual* call for registering the delegate in some function:

    StaticFunctions::addGeoRefUpdateListener(this, &AMyPawn::HandleGeoRefUpdate);

which *internally* just does the `MyGeoRef->GeoRefUpdated.AddDynamic(...);` (so that the `MyGeoRef` is not exposed). This **could** also allow further wiring for the non-default cases - maybe 

    StaticFunctions::createNonDefaultGeoRef("MySpecialGeoRef", someEllipsoid);
    StaticFunctions::addGeoRefUpdateListener("MySpecialGeoRef", someActor, ....);
    StaticFunctions::addGeoRefUpdateListener("MySpecialGeoRef", someOtherActor, ....);
    StaticFunctions::notifyUpdateGeoRef("MySpecialGeoRef"); // Broadcasts update to all registered actors
    someActor->position = StaticFunctions::convertToUE(latLon, "MySpecialGeoRef");
    StaticFunctions::destroyNonDefaultGeoRef("MySpecialGeoRef");
    someOtherActor->position = StaticFunctions::convertToUE(latLon, "MySpecialGeoRef"); // May use the default...?

But ... admittedly, there's a lot of "guessing what **might** work and make sense" involved on my side. I'd like to derive some "actionable items" from that. I'll probably have to read a bit more of the current code, maybe I can come up with some more specific suggestions for refactoring, aiming at things like 1. avoiding the reference/deletion hassle, 2. generalizing the notification, 3. better encapsulating the bounding volume update and such.

**Comment by javagl** [comment]
I've locally been doing some experiments, and tried to get an overview and better understand the current state. Some questions from my local notes (to be elaborated further when necessary - right now, they are really rather internal notes). 

---
 
 - Should sublevels be CONTAINED in GeoRef? (Likely: No)
 - Should SunSky be OWNED by GeoRef (Likely: No)
 
 (In both cases, it may be hard to come up with something better...)
  
 - There are way too many conversion functions in GeoRef...
 - There should hardly be any functions in GeoRef that return a matrix. From the current state of my triage, nearly all places where a GeoRef is used use these matrices just to transform vectors on their own. The GeoRef should just offer the right conversion functions.
 - The functions should be "consistent" regarding the "domain" of their arguments: A function like `doSomething(glm::dvec3, FRotator)` looks wrong. It should be `doSomething(FVector, FRotator)` or `doSomething(glm::dvec3, glm::dvec3)`. 

- Could we describe some "best practice" about ~"How to GeoRef an object?" (I.e. what are the actual, conceptual, important differences between including something in a GeoRef'd sublevel, attaching a "CesiumGeoreferenceComponent" to IT, or attaching IT to a "CesiumGlobeAnchorParent"? And if there are no important ones: Should there be only ONE way to GeoRef something?)

**Comment by xuelongmu** [comment]
> Should SunSky be OWNED by GeoRef (Likely: No)
@nithinp7 opened a PR https://github.com/CesiumGS/cesium-unreal/pull/281 for this a while back that discussed this, I think we can now remove the SunSky reference from Georeference once the interface PR is merged https://github.com/CesiumGS/cesium-unreal/pull/398

I'm also not sure why sublevels are contained in GeoRef, but not sure where else to put them either (perhaps the World Outliner, though I'm not sure if we could override that easily in a plugin).

I agree with all the other points you mentioned.

For how to georef an object, my understanding is that putting a vanilla UE4 actor in a GeoRef'd sublevel mitigates the floating-point inaccuracy of working on a global scale, with the caveat being that the actor must stay within the sublevel's bounds. On the other hand, the GeoreferenceComponent preserves the "true" accuracy of the owning actor regardless of where you are on the globe. Some systems (esp. physics actors) may play nicer by being in a sublevel, as the GeoreferenceComponent will teleport the owning actor to correct its position, which can cause problems with certain Unreal systems (though these potential issues may have been addressed with https://github.com/CesiumGS/cesium-unreal/issues/307)

Having too many GeoreferenceComponents will also impact performance because it runs on Tick, so probably the rule of thumb is: for actors that will be traversing the globe, add a component. Otherwise, put it in a sublevel.

I cannot find any (!) instances where the `CesiumGlobeAnchorParent` is used. We should probably remove it if that's the case.

After thinking more about the blueprint function library, I'm proposing something like this:

```
static ACesiumGeoreference* DefaultGeoref;

// blueprint exposed functions cannot be overloaded. assume default georef for BP functions.
UFUNCTION(BlueprintCallable)
FTransform ConvertToUEFrameUsingDefault() { ConvertToUE(DefaultGeoref); }

FTransform ConvertToUEFrame(ACesiumGeoreference* Georef) { ConvertToUEFrame(Georef->_someGlmMatrix); }

glm::dmat4 ConvertToUEFrame(glm::dmat4 InMatrix);
```

Actors that implement `NotifyGeoreferenceUpdated(ACesiumGeoreference* Georef)` (re: https://github.com/CesiumGS/cesium-unreal/pull/398) will call the first method if they're BP, or if they're C++ then they can choose between the default or a specific georef (examples including AGlobeAwareDefaultPawn, ACesium3DTileset, ACesiumGeoreferenceComponent) that they store their own reference to.

**Comment by nithinp7** [comment]
@javagl With regards to the need for the georef to return matrices, I think this is a good example of its usefulness: https://github.com/CesiumGS/cesium-unreal/blob/c21fd0dd96631b14c4454d0c519899c67d43c7e6/Source/CesiumRuntime/Private/CesiumGeoreferenceComponent.cpp#L368

where `unrealWorldToEcef` was retrieved as a const reference from the georeference. The idea here is that we are nailing down a georeference for an Unreal object so that at any point in the future, it can be properly placed on the globe regardless of what the future state of the georeference is. To do this, we take the current `unrealWorldToEcef` and construct an `actorToEcef` out of it. This `actorToEcef` will be the same regardless of future changes in the georeference, so it is a very useful local matrix to calculate and keep, in fact definitionally it _is_ the georeference of the object. Then in the future, to render the object, we use `actorToEcef` and combine it with the future `ecefToUnrealWorld` to establish a `actorToUnrealWorld` (pretty much a model matrix) for the rendering.

**Comment by javagl** [comment]
@nithinp7 There certainly are places where a matrix is required (even though, for my local experiments and while playing around, I changed the function that you linked to, to use some `computeToECEF` function instead of manually assembling the matrix. But that was *not* a "proposal", but rather a "convenience change" for that particular point, and without understanding everything in all depth).

---

Interlude: This lack of understanding shows up in my difficulties of undertanding the state changes: 

>  This actorToEcef will be the same regardless of future changes in the georeference, 

The `_actorToECEF` is modified in various places (e.g. the `_setECEF` and the `Snap...` functions). It is updated in `_updateActorToECEF` (which is called, under certain conditions, which involve `this->_georeferenced`, of which I'm not sure whether or not it is the same as `this->Georeference != nullptr`). One of the "sources" of these updates is `OnUpdateTransform`. I'll try to break this down to one particular example (even though it is only one of many examples) : 

- Who calls the `Snap` functions? (It's complicated - they are `BlueprintCallable`, but there's also this "auto snap" flag...)
- Who calls `OnUpdateTransform`?
- ---> Will the call to `OnUpdateTransform` **not** overwrite any changes to `_actorToECEF` that may have been done in the `Snap` call?

**tl;dr**: What **is** `_actorToECEF` at any given point in time? 

If we have the choice between 1. doing some matrix multiplication in every tick, even though it could be avoided, and 2. having to understand what's going on there, I'd **strongly** prefer the former. We can think about how to avoid that matrix recomputation **if** we do a profiler run and see that this particular matrix multiplication reduces the number of frames per second from 100 to 99. (And... that won't happen, of course).

---

Regarding the static function library: 

I had sketched **empty** "templates" a while ago, and just added them to my "experiments" branch at https://github.com/CesiumGS/cesium-unreal/commit/3c7d089049084834b15baef35f520b5b93ffbe1a 

I thought that it could make sense to separate the "`double`-versions" and the "Blueprint-Versions". But that's certainly up for debate. The vague reasoning beind that for me was

- Separating between the ones that *solely* operate on `glm` and those that operate on `FVector` could add some clarity
- If UE5 really supports `double` (fingers crossed), we might add one that operates on `DVector`, without any changes to the existing ones
- (For the time of planning and experiments, *before* the result is merged to `master`) : If the functions are separated, then merging them into one class can be done in 30 seconds. If they are merged, separating them into two classes could take hours.

---

Beyond that, the branch does contain some minor cleanups, but not anything "important".

Except, maybe, a *first* attempt to make the `Ellipsoid` a member (instead of always defaulting to  WGS84), via https://github.com/CesiumGS/cesium-unreal/commit/356c3e385f1bf8765b3c9ae634fd743712bbc815 . **But** I stumbled over a problem there: Adding the `UPROPERTY` for the ellipsoid radii messed something up: Afterwards, the deserialization of the matrix arrays did not seem to work any more. They had been initialized properly in the constructor, but then had been "overwritten" with garbage data at some point. The confusing thing is: This happened even when *not* loading an existing, serialized version of the GeoRef, but also when *instantiating* a new one. I have to investigate that further...

---

One detail: I tried around with a `ComponentVisualizer`, in https://github.com/CesiumGS/cesium-unreal/commit/62b8922b00c8f2b5ea9c013bea7e75a4fa183272 , to see whether it could be possible to draw some coordinate systems that indicate orientations of GeoRefs or other things. That *might* come in handy at some point. At least, I *hoped* that it could give me a better understanding of all the coordinate transforms (and maybe some helpful screenshots for the documentation in the long run). But that's also not critical right now.

**Comment by xuelongmu** [comment]
Currently working on the blueprint function library, I am taking your idea of separating out the two classes @javagl and I think it's working nicely.

@nithinp7 I see that `ACesiumGeoreference::GetDefaultForActor` calls `Actor->GetLevel()`. I was wondering what the difference would be between 'Actor->GetLevel()' and `'SomeUObject->GetWorld()`?

There's a nice metadata specifier in blueprint callable functions that can automatically pass in a `UObject`, with the stipulation that we can't call `GetLevel()` on `UObject`. See https://answers.unrealengine.com/questions/838936/how-to-get-worldcontextobject-inside-a-c-bp-functi.html# (second answer)

**Comment by xuelongmu** [comment]
After discussing with @javagl with how to move forward given some experiments in https://github.com/CesiumGS/cesium-unreal/pull/459, https://github.com/CesiumGS/cesium-unreal/tree/bp-function-library-ext, and https://github.com/CesiumGS/cesium-unreal/tree/geo-transforms-draft, we've agreed on a path forward to resolving hopefully many of the issues previously mentioned in this thread and bring some significant UX improvements.

- A new `cesium-native` class (let's call it `GeoTransforms` for now) will be created for storing georeference state.
    - Double-precision conversion functions will be moved from `CesiumGeoreference` to `GeoTransforms`.
    - `GeoTransforms` will hold a property for the ellipsoid type, so that in the future, we can expose configuration of non-WGS 84 ellipsoids.
    - Creating a new `CesiumGeoreference` instance will (in most cases) create a new `GeoTransforms` analog
    - `CesiumGeoreference` will act as a way of interfacing with and modifying `GeoTransforms` from Unreal land. It will not hold any state on its own. The (inaccurate) blueprint conversion functions will remain there, but probably renamed without the "inaccurate" flag. Comments will direct users to double-precision equivalents when that is required.
    - The lifecycle of `GeoTransforms` will be handled outside of Unreal's UObject garbage collection, which has a key advantage: when a `CesiumGeoreferenceDefault` is deleted, the native `GeoTransforms` analog is not, and when a new `CesiumGeoreferenceDefault` is created (e.g. from calling `GetDefaultForActor`), it will look for the default `GeoTransforms` and use that. Therefore, the state will not be erased when accidentally or intentionally deleting the designated default georeference.
- We will not be continuing with the static function library idea https://github.com/CesiumGS/cesium-unreal/issues/318, as it was difficult to keep track of state and find/provide the correct input matrices for a given conversion. The conversions functions will remain non-static.
- `Bounding Volume Center` will be deprecated, as there is little use case and is currently broken with the CesiumSunSky system. True Origin will remain and we will fix the placement issue described in https://github.com/CesiumGS/cesium-unreal/issues/483. 
- Georeference update notifications will be handled via Unreal delegates. (Tentative: the signature of the delegate callback functions will be a `CesiumGeoreference` pointer, which can be used for subsequent calculations) We will also add a delegate to notify when the georeference will be deleted so that actors can react accordingly if needed. The combination of this and the `Bounding Volume Center` deprecation means that `ICesiumGeoreferenceable` will be deprecated and removed entirely. This will also close https://github.com/CesiumGS/cesium-unreal/issues/321 and https://github.com/CesiumGS/cesium-unreal/pull/398 as they are no longer needed.

Outstanding issues:
- https://github.com/CesiumGS/cesium-unreal/issues/470 Sublevels cannot be removed, we probably just have to remove a sublevel from the struct array if there is no corresponding .umap asset 
- GeoRef instances that are created via C++ are not always added in the World Outliner
- Georef instances that are removed at edit-time are not garbage-collected. We need a robust way of finding a valid default georeference in this case, as the name "DefaultGeoreference" might be taken by a actor marked `Pending Kill`. (One interesting point: we may be able to [destroy objects manually](https://nerivec.github.io/old-ue4-wiki/pages/garbage-collection-dynamic-memory-allocation.html#Destroying_Objects) by calling `ConditionalBeginDestroy()` and free up that name)
- https://github.com/CesiumGS/cesium-unreal/issues/422 Race condition when moving GlobeAwareDefaultPawn + actor with CesiumGeoreferenceComponent on same tick (could be due to us calling `SetNewWorldOrigin()` on rebase rather than `RequestNewWorldOrigin()`, which IIRC waits until the following tick to perform updates, perhaps for this very reason)

Next steps:
- Merge `GeoTransform` class from `geo-transforms-draft` into `bp-function-library` branch
- Merge ellipsoid handling and other refactoring from `bp-function-library-ext` into `bp-function-library` branch
- Cherry-pick `GeoreferenceComponent` refactoring from https://github.com/CesiumGS/cesium-unreal/commit/695f09e8b7d3936c0409b6c941daa20ed2a6edf4 into `bp-function-library` branch

**Comment by javagl** [comment]
Coming back to an [earlier comment](https://github.com/CesiumGS/cesium-unreal/pull/459#issuecomment-850692242) : I noticed that there are several functions that receive the "world origin" and a "position". In these cases, the "position" is usually given *relative* to the world origin. 

This makes sense for the Unreal world, where these positions have to be maintained separately, due to precision issues. But I think that we currently have several places with functions like

    dvec3 doSomething(dvec3 origin, dvec3 relativePosition) {
        ...
       return someMatrix * (origin + relativePosition);
    }

that *could* be simplified to

    dvec3 doSomething(dvec3 absolutePosition) {
        ...
       return someMatrix * relativePosition;
    }

The current conversion functions (or rather: the ones that result from the current refactoring pass) should be reviewed for this pattern.

**Comment by javagl** [comment]
I made a first pass on https://github.com/CesiumGS/cesium-unreal/commits/bp-function-library . Many of the commits are smaller fixes. Significant changes are:

- Renamed "Ue" to "Unreal" in function names, via https://github.com/CesiumGS/cesium-unreal/commit/e5ebaf2335fe3e3516a54503b35be3517435903e , similar to the function names in the BP library. If requested, I can undo that, but we'll have to sort out the precise-vs-`Inaccurate` function naming anyhow..
- The functions to convert "rotators" did not yet have a purely glm-based counterpart (and their implementation in the latest states had been commented out). These are now using `glm` as of https://github.com/CesiumGS/cesium-unreal/commit/c01273e09c8b6b74cfab28a9183858f9ff5a3778
- Where I was entirely not sure: I tried to update the GeoRefComponent to consistenly use the conversion functions that are offered by the GeoRef - also to avoid using "the wrong" ellipsoid. Two tasks are colliding there: 1. using the GeoRef consistently, and 2. figuring out where and when the GeoRef is assigned to the component. There are several TODOs for a review, via https://github.com/CesiumGS/cesium-unreal/commit/edb24e80e87b2816b8072136ef2d548c9f1eb7fe 
- (An otherwise unrelated fix that was part of another PR: Make sure that the subleves match the ones that are actually present, via https://github.com/CesiumGS/cesium-unreal/commit/8791c80f9784c90b7121a021e0b84d668db31ab9 )
- Tried to update the `GetDefaultForActor` as discussed: It is called `GetDefaultGeoreference` now, receives this magic "WorldContextObject", and uses tags to identify the ("global") default GeoRef, via https://github.com/CesiumGS/cesium-unreal/commit/43c943591c95f7142ddd61ef0e4b0c1d1213b746 - This seems to work quite well, and fixes **half** of https://github.com/CesiumGS/cesium-unreal/issues/242 (the other half is related to the CreditSystem, where a similar pattern could be used)
- Added a delegate for the updates via https://github.com/CesiumGS/cesium-unreal/commit/3565d1d8385678dd6a0f6ea9c6c10fee40b1c892 - Right now, this is *only* `GeoreferenceUpdated`, and *only* used for the tileset root. But it also seems to work pretty well, and could IMHO replace the `IGeoreferenceListener` completely.

There are many further TODOs (some smaller ones, like removing obsolete functions or classes), but I also started to test (or rather: try out) the results in Blueprints and in general. (An aside: Right now, the dynamic pawn is initialized to a wrong position when starting to play. I hope that I'll catch the reason for this during the review+cleanups).

**Comment by javagl** [comment]
I think I already mentioned that somewhere as a side note, but we should really (I mean **REALLY**) try to be consistent with our function naming.  People should have have to wonder (or helplessly try to look up) whether "ECEF" is the same as "EllipsoidCentered".

**Comment by xuelongmu** [comment]
> ```
> dvec3 doSomething(dvec3 origin, dvec3 relativePosition) {
>     ...
>    return someMatrix * (origin + relativePosition);
> }
> ```
> 
> that _could_ be simplified to
> 
> ```
> dvec3 doSomething(dvec3 absolutePosition) {
>     ...
>    return someMatrix * absolutePosition;
> }
> ```
> 

> 
> EDIT: A strongly related question: What are cases where someone wants to perform a computation that involves the "world origin", where the actual world origin should _not_ be the one of the GeoRef? 

I was doing some testing earlier and trying to see if there's some scenario like this. My guess would be _maybe_ this would come up if the user is performing the computation from an actor in a sublevel, and the georef is in the persistent level. But have to keep digging into this.

**Comment by xuelongmu** [comment]
@javagl From doing some testing today, I think I can say with reasonable confidence that the computations involving the world origin can use the world origin of the Georef. More broadly speaking, I think that there can only be one floating world origin set at any given time, so no matter which actor you call `GetWorld()->OriginLocation` from, it will return the same value. 

Some interesting debug findings: even when an actor is inside a sublevel, calling `GetCurrentLevelName()` will return the name of the "overall" map. So in my test project, I have a WorldMap that holds a Miami sublevel, and a test actor in the Miami sublevel only. However, when calling `GetCurrentLevelName()` in BP, it returns `UEDPIE_0_WorldMap`. On the other hand, the Georeference (in C++) returns `UEDPIE_0_WorldMap` when calling `GetWorld()->GetName` and `GetWorld()->GetMapName`, but `PersistentLevel` when calling `GetLevel()->GetName()` and `GetWorld()->GetCurrentLevel()->GetName()`. None of my map assets are named `PersistentLevel`, so I assume UE generates this object to represent the persistent level.

The conceptual difference between a World and a Level in UE4 is not something I'm exactly clear on. I will try looking more into this, but I think the earlier conclusion still stands:
>The internal conversion functions could receive this origin as an argument. But the other functions (from GeoRef) would not need that, and could always just pass the world origin of the GeoRef to the low-level functions)
